core/policy/policy_repository.go

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (C) 2019 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package policy
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/mysteriumnetwork/node/market"
+	"github.com/mysteriumnetwork/node/requests"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+)
+
+// PolicyRepository represents async policy fetcher from TrustOracle
+type PolicyRepository struct {
+	client *requests.HTTPClient
+
+	policyURL   string
+	policyLock  sync.Mutex
+	policyRules map[market.AccessPolicy]market.AccessPolicyRuleSet
+
+	fetchInterval time.Duration
+	fetchShutdown chan bool
+}
+
+// NewPolicyRepository create instance of policy repository
+func NewPolicyRepository(client *requests.HTTPClient, policyURL string, interval time.Duration) *PolicyRepository {
+	return &PolicyRepository{
+		client:        client,
+		policyURL:     policyURL,
+		policyRules:   make(map[market.AccessPolicy]market.AccessPolicyRuleSet),
+		fetchInterval: interval,
+	}
+}
+
+// Start begins fetching proposals to repository
+func (pr *PolicyRepository) Start() {
+	pr.fetchShutdown = make(chan bool, 1)
+	go pr.fetchLoop()
+}
+
+// Stop ends fetching proposals to repository
+func (pr *PolicyRepository) Stop() {
+	pr.fetchShutdown <- true
+}
+
+// Policy converts given value to valid policy rule
+func (pr *PolicyRepository) Policy(policyID string) market.AccessPolicy {
+	return market.AccessPolicy{
+		ID:     policyID,
+		Source: fmt.Sprintf("%v%v", pr.policyURL, policyID),
+	}
+}
+
+// Policies converts given values to list of valid policies
+func (pr *PolicyRepository) Policies(policyIDs []string) []market.AccessPolicy {
+	policies := make([]market.AccessPolicy, len(policyIDs))
+	for i, policyID := range policyIDs {
+		policies[i] = pr.Policy(policyID)
+	}
+	return policies
+}
+
+// AddPolicies adds given policy to repository. Also syncs policy rules from TrustOracle
+func (pr *PolicyRepository) AddPolicies(policies []market.AccessPolicy) error {
+	pr.policyLock.Lock()
+	policyRulesNew := pr.policyRules
```

Comment:
`policyRules` is a map. Assigning it to another variable not copy data, just assign the reference. It does not protect the map from a race condition.

Reply:
What do do then? I don't want to lock reads of this variable for time spent on fetching data to this variables

Reply:
1. Lock pr.policyRules to read  source ids into string array etc. since fetchPolicyRules could accept source only.
2. Now iterate all sources and fetch into new map which don't need any lock.
3. Now lock again and remap from new map into your destination.

Reply:
1. fetchPolicyRules could accept source instead of whole policy.
2. Now you can lock and read into source array and unlock.
3. Now iterate all sources and fetch into new map which don't need to be locked.
4. Now lock again and merge new map values into policyRules. Unlock. Done.

Reply:
Anyways, I've just remade to array instead of map

Reply:
Your latest changes didn't address @soffokl 's comment. You are copying reference, but releasing the lock, allowing race conditions with other `AddPolicies` calls.

You don't need to lock reads at all, only in `AddPolicies`:
```go
pr.policyLock.Lock()
defer pr.policyLock.Unlock()
policyListNew := deep copy pr.policyList
update policyListNew
pr.policyList = policyListNew
return nil
```
This will now allow concurrent `AddPolicies`, but will allow concurrent reads at any point: before new reference is updated on `pr`, all reads will return current access policies.

Reply:
Thanks on ideas. Improved locking of list